queasars/job_shop_scheduling/domain_wall_hamiltonian_encoder.py

```diff
@@ -0,0 +1,528 @@
+        """
+        if not self._encoding_prepared:
+            self._prepare_encoding()
+
+        if not self._hamiltonian_prepared:
+            self._prepare_hamiltonian()
+
+        return SparsePauliOp.sum(self._local_terms)
+
+    def translate_result_bitstring(self, bitstring: str) -> JobShopSchedulingResult:
+        """
+        Translates a bitstring as measured on a quantum circuit to it's corresponding job shop scheduling result.
+        The bitstring must only contain 1 and 0s and must be of length self.n_qubits
+
+        :arg bitstring: string of 1 and 0s measured on the quantum circuit
+        :type bitstring: str
+        :return: the JobShopSchedulingResult this bitstring represents
+        """
+        if len(bitstring) != self.n_qubits:
+            raise ValueError("The bitstring length does not match the problem size!")
+
+        def translate(string: str) -> int:
+            if string == "1":
+                return 1
+            if string == "0":
+                return 0
+            raise ValueError("The bitstring may not contain any value apart from 1 or 0!")
+
+        bit_list = [translate(char) for char in bitstring]
+
+        job_schedules: dict[Job, tuple[PotentiallyScheduledOperation, ...]] = {}
+        for job in self.jssp_instance.jobs:
+            scheduled_operations: list[PotentiallyScheduledOperation] = []
+            for operation in job.operations:
+                domain_wall_variable = self._operation_start_variables[operation]
+                start_time = domain_wall_variable.value_from_bitlist(bit_list=bit_list)
+                if start_time is not None:
+                    scheduled_operations.append(ScheduledOperation(operation=operation, start_time=start_time))
+                else:
+                    scheduled_operations.append(UnscheduledOperation(operation=operation))
+            job_schedules[job] = tuple(scheduled_operations)
+
+        return JobShopSchedulingResult(problem_instance=self.jssp_instance, schedule=job_schedules)
+
+    def _prepare_encoding(self) -> None:
+        """Counts the needed qubits to encode the problem" and assigns the necessary domain wall variables"""
+        for job in self.jssp_instance.jobs:
+            for i, operation in enumerate(job.operations):
+                if operation.machine not in self._machine_operations:
+                    self._machine_operations[operation.machine] = []
+                self._machine_operations[operation.machine].append(operation)
+
+                start_offset = sum(operation.processing_duration for j, operation in enumerate(job.operations) if j < i)
+                end_offset = sum(operation.processing_duration for j, operation in enumerate(job.operations) if j >= i)
+
+                n_start_times = self.time_limit - (start_offset + end_offset) + 1
+
+                if n_start_times < 1:
+                    raise ValueError("There is no feasible solution for the given time_limit!")
```

Comment:
Maybe add the relevant debugging information.

Reply:
I added some relevant debugging information.